blockdevice/probe/probe.go

```
@@ -34,106 +26,96 @@ type ProbedBlockDevice struct {
-	// and it will fail to open.
-	err = retry.Constant(5*time.Second, retry.WithUnits((50 * time.Millisecond))).Retry(func() error {
-		if f, err = os.OpenFile(path, os.O_RDONLY|unix.O_CLOEXEC, os.ModeDevice); err != nil {
-			if os.IsNotExist(err) {
-				return retry.ExpectedError(err)
-			}
-
-			return retry.UnexpectedError(err)
-		}
+// All probes a block device's file system for the given label.
+func All(options ...SelectOption) (all []*ProbedBlockDevice, err error) {
+	var infos []os.FileInfo
 
-		return nil
-	})
-	if err != nil {
+	if infos, err = ioutil.ReadDir("/sys/block"); err != nil {
 		return nil, err
 	}
 
-	//nolint: errcheck
-	defer f.Close()
+	for _, info := range infos {
+		devpath := "/dev/" + info.Name()
 
-	superblocks := []filesystem.SuperBlocker{
-		&iso9660.SuperBlock{},
-		&vfat.SuperBlock{},
-		&xfs.SuperBlock{},
-	}
+		probed := probePartitions(devpath)
 
-	for _, sb := range superblocks {
-		if _, err = f.Seek(sb.Offset(), io.SeekStart); err != nil {
-			return nil, err
-		}
+		for _, part := range probed {
+			add := true
+			for _, matches := range options {
+				add, err = matches(part)
+				if err != nil {
+					return nil, err
+				}
 
-		err = binary.Read(f, binary.BigEndian, sb)
-		if err != nil {
-			return nil, err
-		}
+				if !add {
+					break
+				}
+			}
 
-		if sb.Is() {
-			return sb, nil
+			if add {
+				all = append(all, probed...)
+			}
 		}
 	}
 
-	return nil, nil
+	return all, nil
 }
 
 // GetDevWithFileSystemLabel probes all known block device's file systems for
 // the given label.
 func GetDevWithFileSystemLabel(value string) (probe *ProbedBlockDevice, err error) {
 	var probed []*ProbedBlockDevice
 
-	if probed, err = All(); err != nil {
+	if probed, err = All(WithPartitionsLabeled(value)); err != nil {
 		return nil, err
 	}
 
-	return filterByLabel(probed, value)
+	if len(probed) == 0 {
+		return nil, os.ErrNotExist
+	}
+
+	return probed[0], nil
```

should we return an error if several partitions are found?

---

yeah, let's do it